devito/finite_differences/derivative.py

```diff
@@ -277,8 +285,25 @@ def evaluate(self):
         # types of discretizations.
         return self._eval_fd(self.expr)
 
+    @property
+    def _eval_deriv(self):
+        return self._eval_fd(self.expr)
+
     def _eval_fd(self, expr):
-        expr = getattr(expr, 'evaluate', expr)
+        """
+        valuate finite difference approximation of the Derivative.
+        The evaluation goes in four steps:
+        - 1: evaluate derivatives within the expression. For example `f.dx * g` will
+        evaluate `f.dx` first.
+        - 2: Evaluate the finite difference for the (new) expression
+        - 3: Evaluate remaining (as `g` may need to be evaluated at a different point)
```

Comment:
for the sake of speed, we need to make sure that no useless xreplaces are applied if no substitutions are needed (might be already this way, just pointing it out)